Rework ReplicationLagWarning alert

williamlardier opened the pull request:
The previous version was creating alerts after a fresh install:
- The oplog would not be full in this case
- No operation might cause the lag to be wrongly high, or
- ...the platform is in a transient / startup phase

We would create this alert even if the oplog was not full, because the delta was too small. We here avoid that by ensuring this alert is only created if the oplog circular buffer is fully utilized. This makes more sense because in other cases, it means we have margin.

The strategy here is to use the oplog size as reported by mongodb PRIMARY and detect if the usage is reaching the max. We start the alert at 95% to keep a small margin in case of heavy workloads.

Issue: ZENKO-5047

Seragonia committed:
Rework ReplicationLagWarning alert

The previous version was creating alerts after a fresh install:
- The oplog would not be full in this case
- No operation might cause the lag to be wrongly high, or
- ...the platform is in a transient / startup phase

We would create this alert even if the oplog was not full, because
the delta was too small. We here avoid that by ensuring this
alert is only created if the oplog circular buffer is fully
utilized. This makes more sense because in other cases, it means
we have margin.

Issue: ZENKO-5047

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/zenko/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-5047 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.11.12`

* `2.12.5`

* `2.13.0`


Please check the `Fix Version/s` of ZENKO-5047, or the target
branch of this pull request.

welansari commented on monitoring/mongodb/alerts.yaml:
```diff
@@ -103,24 +103,38 @@ groups:
+        )
+        AND on(rs_nm)
+        (
+          sum by (rs_nm) (
+            (
+              mongodb_oplog_stats_storageStats_size{namespace="${namespace}", pod=~"${service}.*"}
+              / 
+              mongodb_oplog_stats_storageStats_maxSize{namespace="${namespace}", pod=~"${service}.*"}
             )
-          )
-        ) * 100 > ${replicationLagWarningThreshold},
+            * on(pod) group_left()
+            (mongodb_mongod_replset_my_state{namespace="${namespace}", pod=~"${service}.*"} == 1)
+          ) >= 0.95
+        ),
```
We don't necessarily have to have a fully utilized oplog buffer to encounter replication issues right ? In that case, this fixes the issue with false positives at init but might hide lag issues happening early after the creation of the platform. The oplog size is configurable (not sure if we have customers with a larger size than the default) which means that it can be fairly large and can take a while before it's full, which will make the alert trigger much later than it's supposed to.

williamlardier replied:
> We don't necessarily have to have a fully utilized oplog buffer to encounter replication issues right ?

No indeed, but product wise, we do not want to alert if there is no risk. The replication lag can be higher during this period with no issue for customers. When it reaches 95% (not the max), then we can start having problems. If a platform was already lagging before the oplog reaches the max storage, then it will need a resync anyway, it's not a replication lag issue anymore.

The oplog size is indeed configurable although we don't change it, so it's 5% (minimum 990MB) of the available storage.

williamlardier replied:
A trade-off could be to choose a lower threshold, like 50%, leaving enough room for the startup phases, and some operations. What do you think?

welansari replied:
I think that works yes, LGTM otherwise

welansari replied:
Maybe we could just make the threshold configurable ?

Seragonia committed:
Set oplog size threshold as 50%

This will still ensure we don't fire the alert for nothing, and
that the alert can fire when the platform had some events alrady.

Also make the threshold configurable.

Issue: ZENKO-5047

ghost commented:
/approve

ghost commented:
/after_pull_request 2256

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/ZENKO-5047` into
`development/2.11`
* [w/2.12/improvement/ZENKO-5047](https://github.com/scality/zenko/compare/development/2.12...w/2.12/improvement/ZENKO-5047) will be merged into `development/2.12` 
* [w/2.13/improvement/ZENKO-5047](https://github.com/scality/zenko/compare/development/2.13...w/2.13/improvement/ZENKO-5047) will be merged into `development/2.13` 



The following branches will **NOT** be impacted:

* `development/2.10`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Build failed



The [build](https://github.com/scality/Zenko/actions?query=branch%3Aw/2.12/improvement/ZENKO-5047) for [commit](https://github.com/scality/Zenko/commit/7cdf15f57fc817056daef941a706f18566b09a4b) did not succeed in branch [w/2.12/improvement/ZENKO-5047](https://github.com/scality/zenko/compare/development/2.12...w/2.12/improvement/ZENKO-5047)



*The following options are set:* **approve**

ghost commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/ZENKO-5047` into
`development/2.11`
* [w/2.12/improvement/ZENKO-5047](https://github.com/scality/zenko/compare/development/2.12...w/2.12/improvement/ZENKO-5047) will be merged into `development/2.12` 
* [w/2.13/improvement/ZENKO-5047](https://github.com/scality/zenko/compare/development/2.13...w/2.13/improvement/ZENKO-5047) will be merged into `development/2.13` 



The following branches will **NOT** be impacted:

* `development/2.10`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Build failed



The [build](https://github.com/scality/Zenko/actions?query=branch%3Aw/2.12/improvement/ZENKO-5047) for [commit](https://github.com/scality/Zenko/commit/0770a0ed5b2151370c454ecc4269435934f532aa) did not succeed in branch [w/2.12/improvement/ZENKO-5047](https://github.com/scality/zenko/compare/development/2.12...w/2.12/improvement/ZENKO-5047)



*The following options are set:* **approve**

bert-e commented:
# Build failed



The [build](https://github.com/scality/Zenko/actions?query=branch%3Aw/2.13/improvement/ZENKO-5047) for [commit](https://github.com/scality/Zenko/commit/64110eb5566049a92348dd724dbaba8eefe93f31) did not succeed in branch [w/2.13/improvement/ZENKO-5047](https://github.com/scality/zenko/compare/development/2.13...w/2.13/improvement/ZENKO-5047)



*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.11`

* :heavy_check_mark: `development/2.12`

* :heavy_check_mark: `development/2.13`



The following branches will **NOT** be impacted:

* `development/2.10`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.11`

* :heavy_check_mark: `development/2.12`

* :heavy_check_mark: `development/2.13`



The following branches have **NOT** changed:

* `development/2.10`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue ZENKO-5047.

Goodbye williamlardier.
